### PR TITLE
enterprise license support

### PR DESCRIPTION
**Pull request by kvahed**
### Scope & Purpose

*Mostly error codes for Enterprise Feature*

- [ ] :hankey: Bugfix (requires CHANGELOG entry)
- [X] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [X] :book: CHANGELOG entry made

#### Backports:

- [ ] No backports required
- [ ] Backports required for: *(Please specify versions and link PRs)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [X] Docs PR: will announce 
- [X] Enterprise PR:
- [ ] GitHub issue / Jira ticket number:
- [ ] Design document: 

### Testing & Verification

*(Please pick either of the following options)*

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [X] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [X] This PR adds tests that were used to verify all changes:
  - [X] Added new C++ **Unit tests**
  - [X] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

Link to Jenkins PR run:

### Documentation

> All new features should be accompanied by corresponding documentation. 
> Bugs and features should furthermore be documented in the CHANGELOG so that
> support, end users, and other developers have a concise overview. 

- [ ] Added entry to *Release Notes* 
- [ ] Added a new section in the *Manual* 
- [ ] Added a new section in the *HTTP API* 
- [ ] Added *Swagger examples* for the HTTP API  
- [ ] Updated license information in *LICENSES-OTHER-COMPONENTS.md* for 3rd party libraries

### External contributors / CLA Note 

Please note that for legal reasons we require you to sign the [Contributor Agreement](https://www.arangodb.com/documents/cla.pdf)
before we can accept your pull requests.

### CODE REVIEW

File: arangod/Cluster/ServerState.cpp [hunk 1]
```diff
@@ -319,14 +319,26 @@ ServerState::Mode ServerState::setServerMode(ServerState::Mode value) {
...
+    _serverstate_readonly.exchange(false, std::memory_order_release);
+  } else if (ro == API_TRUE) {
+    _serverstate_readonly.exchange(true, std::memory_order_release);
+  } else if (ro == LICENSE_FALSE) {
+    _license_readonly.exchange(false, std::memory_order_release);
+  } else if (ro == LICENSE_TRUE) {
+    _license_readonly.exchange(true, std::memory_order_release);
+  }
+  return _serverstate_readonly.load(std::memory_order_acquire) ||
+    _license_readonly.load(std::memory_order_acquire);
```
Comment (jsteemann) [inline arangod/Cluster/ServerState.cpp:341]: This is a behavior change: the previous implementation returned the _previous_ value of the read-only flag, the new implementation returns the new value.
I think it makes sense to restore the old behavior so everything remains fully downwards-compatible.

File: arangod/Cluster/ServerState.h [hunk 2]
```diff
@@ -57,6 +57,13 @@ class ServerState {
     STATE_SHUTDOWN        // used by all roles
   };
 
+  enum ReadOnlyMode {
```
Comment (jsteemann) [inline arangod/Cluster/ServerState.h:60]: I would prefer if we could turn this into an `enum class`. That way we will not have any implicit conversions and only the defined enum values can be used.

**Review by jsteemann — CHANGES_REQUESTED** [review]

File: arangod/GeneralServer/GeneralServerFeature.cpp [hunk 3]
```diff
@@ -653,8 +654,10 @@ void GeneralServerFeature::defineHandlers() {
 #ifdef USE_ENTERPRISE
   if (backup.isAPIEnabled()) {
     _handlerFactory->addPrefixHandler("/_admin/backup",
-                                    RestHandlerCreator<arangodb::RestHotBackupHandler>::createNoData);
+                                      RestHandlerCreator<arangodb::RestHotBackupHandler>::createNoData);
   }
+  _handlerFactory->addPrefixHandler("/_admin/license",
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/GeneralServer/GeneralServerFeature.cpp:659]: It does not. It is inside an ifdef for enterprise builds.

File: arangod/Agency/Store.cpp [hunk 4]
```diff
@@ -668,7 +668,7 @@ bool Store::read(VPackSlice const& query, Builder& ret) const {
   std::vector<std::string> query_strs;
   for (auto const& sub_query : VPackArrayIterator(query)) {
     query_strs.emplace_back(sub_query.copyString());
-    showHidden |= (query_strs.back().find("/.") != std::string::npos);
+    showHidden |= (query_strs.back().find('.') == 0 || (query_strs.back().find("/.") != std::string::npos));
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/Agency/Store.cpp:671]: this is ok. we need to be able to spot ".agency" as well as "/.agency".

File: js/client/modules/@arangodb/testsuites/license.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+// //////////////////////////////////////////////////////////////////////////////
+// / @brief Shared conf
+// //////////////////////////////////////////////////////////////////////////////
+
+exports.setup = function(testFns, defaultFns, opts, fnDocs, optionsDoc, allTestPaths) {
+  Object.assign(allTestPaths, testPaths);
+  // just a convenience wrapper for the regular tests
+  testFns['license'] = ['license-api'];
+
+  // turn off ldap tests by default.
```
Comment (jsteemann) [inline js/client/modules/@arangodb/testsuites/license.js:60]:
```suggestion
  // turn off license tests by default.
```

**Review by jsteemann — APPROVED** [review]

**Review by jsteemann — APPROVED** [review]

**Pull request merged**